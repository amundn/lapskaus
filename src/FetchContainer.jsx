import React from 'react'
import PropTypes from 'prop-types'
import config from './config'

function increment(state) {
  return { loading: state.loading + 1 }
}
function decrement(state) {
  return { loading: state.loading - 1 }
}

function checkStatus(r) {
  if (r.status !== 200) {
    throw new Error(r.statusText + ' (HTTP ' + r.status + ')')
  }
  return r
}

class FetchContainer extends React.Component {
  state = { loading: 0 }

  render() {
    const childrenWithProps = React.Children.map(this.props.children, child =>
      React.cloneElement(child, {
        isLoading: this.state.loading > 0,
        message: this.state.message
      })
    )

    return (
      <div>
        <div>{childrenWithProps}</div>
      </div>
    )
  }

  static propTypes = {
    isLoading: PropTypes.bool,
    message: PropTypes.string
  }
  static childContextTypes = {
    fetchJson: PropTypes.func,
    fetchImage: PropTypes.func
  }

  getChildContext() {
    return {
      fetchJson: this.handleFetchJson,
      fetchImage: this.handleFetchImage
    }
  }

  handleFetchImage = (description, url, callback) => {
    let img = new Image() // TODO: Error handling
    img.onload = function() {
      callback(img)
    }
    img.src = url
  }

  handleFetchJson = (description, url, callback) => {
    if (this.props.fetchJson)
      return this.props.fetchJson(description, url, callback)
    this.handleFetch(description, url, response => {
      response.json().then(json => callback(json))
    })
  }

  handleFetch = (description, url, callback) => {
    console.error(url)
    this.flashMessage(`Loading ${description}...`)
    this.setState(increment)
    url = config.apiUrl + url
    fetch(url)
      .then(checkStatus)
      .then(response => {
        if (!response.ok) throw new Error(response)
        callback(response)
        this.setState(decrement)
      })
      .catch(error => {
        const message = description + ': ' + error.message
        console.error(message)
        this.setState(decrement)
        this.flashMessage(message)
      })
  }

  flashMessage(message) {
    this.setState({ message })
    clearTimeout(this.messageTimer)
    this.messageTimer = window.setTimeout(
      () => this.setState({ message: null }),
      3500
    )
  }
}

export default FetchContainer
