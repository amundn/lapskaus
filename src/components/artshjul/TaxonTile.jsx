import React from 'react'
import { GridListTile, GridListTileBar } from 'material-ui/GridList'
import IconButton from 'material-ui/IconButton'
import OpenInBrowser from 'material-ui-icons/OpenInBrowser'
import config from '../../config'
import { withStyles } from 'material-ui/styles'
import PropTypes from 'prop-types'

const styles = theme => ({
  title: {},
  titleBar: {
    background:
      'linear-gradient(to top, rgba(0,0,0,0.7) 0%, rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)'
  },
  img: {
    transition: 'opacity 0.3s',
    maxWidth: '100%',
    maxHeight: '100%',
    verticalAlign: 'middle',
    margin: 'auto',
    display: 'block'
  }
})

class TaxonTile extends React.Component {
  constructor() {
    super()
    this.state = { opacity: 0 }
  }
  render() {
    const { tile, classes } = this.props
    return (
      <GridListTile
        cols={1}
        rows={1}
        onClick={() => this.props.onGotoTaxon(tile.id)}
        style={{
          margin: '2px'
        }}
      >
        <div
          style={{
            height: '379px',
            width: '379px',
            backgroundImage:
              'url(' + config.apiUrl + 'taxonphoto/' + tile.s + '.jpg)',
            position: 'relative',
            backgroundSize: 'cover',
            backgroundRepeat: 'no-repeat',
            backgroundPosition: 'center center'
          }}
        />
        <GridListTileBar
          title={<span style={{ textTransform: 'capitalize' }}>{tile.p}</span>}
          subtitle={<b>{tile.s}</b>}
          titlePosition="top"
          classes={{
            root: classes.titleBar,
            title: classes.title
          }}
          actionIcon={
            <IconButton
              onClick={e => {
                alert(this.props.tile.s)
                e.stopPropagation()
              }}
            >
              <OpenInBrowser color="white" />
            </IconButton>
          }
        />
      </GridListTile>
    )
  }
}

TaxonTile.propTypes = {
  classes: PropTypes.object.isRequired
}
export default withStyles(styles)(TaxonTile)
