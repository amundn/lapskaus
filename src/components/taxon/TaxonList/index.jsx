import TaxonList from './TaxonList'

export default TaxonList
