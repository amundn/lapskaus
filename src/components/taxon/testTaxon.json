{
  "me": {
    "id": 1281,
    "s": "Magnoliophyta",
    "p": "dekkfroete blomsterplanter"
  },
  "children": [
    {
      "id": 1282,
      "s": "Monocots",
      "p": "enfrobladete blomsterplanter"
    },
    {
      "id": 1295,
      "s": "Eudicots",
      "p": "tofrobladete blomsterplanter"
    }
  ],
  "parent": {
    "id": 1131,
    "s": "Plantae",
    "p": "planteriket"
  },
  "breadcrumb": [
    {
      "id": 0,
      "s": "Life",
      "p": "Life"
    },
    {
      "id": 1131,
      "s": "Plantae",
      "p": "planteriket"
    }
  ]
}