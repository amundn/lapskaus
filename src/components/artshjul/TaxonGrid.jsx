import React from 'react'
import { GridList, GridListTile } from 'material-ui/GridList'
//import OpenInBrowser from 'material-ui-icons/OpenInBrowser'
import Subheader from 'material-ui/List/ListSubheader'
import { withStyles } from 'material-ui/styles'
import PropTypes from 'prop-types'
import TaxonTile from './TaxonTile'

const styles = theme => ({
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-around',
    overflow: 'hidden',
    background: theme.palette.background.paper
  },
  gridList: {
    width: '100%',
    height: '100%'
  }
})

class TaxonGrid extends React.Component {
  render() {
    const classes = this.props.classes
    return (
      <div className={classes.container}>
        <GridList cellHeight={180} className={classes.gridList} cols={3}>
          <GridListTile cols={3} style={{ height: 'auto' }}>
            <Subheader>{this.props.headerText}</Subheader>
          </GridListTile>
          {this.props.tilesData.map(tile => (
            <TaxonTile
              classes
              onGotoTaxon={this.props.onGotoTaxon}
              tile={tile}
              key={tile.id}
            />
          ))}
        </GridList>
      </div>
    )
  }
}

TaxonGrid.propTypes = {
  headerText: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired,
  onGotoTaxon: PropTypes.func
}

export default withStyles(styles)(TaxonGrid)
