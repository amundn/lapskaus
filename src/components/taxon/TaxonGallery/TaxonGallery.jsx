import React from 'react'
import List, { ListItem, ListItemText } from 'material-ui/List'
import Avatar from 'material-ui/Avatar'
import TaxonGrid from './TaxonGrid'
import config from '../../../config'

class TaxonGallery extends React.Component {
  render() {
    const tree = this.props.tree
    if (!tree) return null
    return (
      <div>
        {tree.breadcrumb && (
          <List>
            <Breadcrumb
              onNavigateToTaxon={this.handleGotoTaxon}
              source={tree.breadcrumb}
            />
          </List>
        )}

        <TaxonGrid
          headerText={label(tree.me)}
          tilesData={tree.children}
          onNavigateToTaxon={this.handleGotoTaxon}
          onPickTaxon={this.props.onPickTaxon}
        />
      </div>
    )
  }
}

const label = node => (node.p ? node.p + ' (' + node.s + ')' : node.s)

const Item = ({ node, rankNo, onNavigateToTaxon }) => {
  const text = label(node)
  return (
    <ListItem
      button
      style={{
        xmarginLeft: Math.max(0, rankNo - 2) * 10
      }}
      id={node.id}
      onClick={() => onNavigateToTaxon(node.id)}
    >
      {node.id > 0 && (
        <Avatar src={config.apiUrl + 'taxonPhoto/' + node.s + '.jpg'} />
      )}
      <ListItemText
        primary={
          <span
            style={{
              textTransform: 'capitalize'
            }}
          >
            {text}
          </span>
        }
      />
    </ListItem>
  )
}

const Breadcrumb = ({ source, onNavigateToTaxon }) => (
  <div>
    {source.map((s, idx) => (
      <Item
        rankNo={idx + 1}
        key={s.id}
        node={s}
        onNavigateToTaxon={() => onNavigateToTaxon(s.id)}
      />
    ))}
  </div>
)

export default TaxonGallery
