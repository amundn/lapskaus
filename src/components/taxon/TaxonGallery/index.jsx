import TaxonGallery from './TaxonGallery'

export default TaxonGallery
