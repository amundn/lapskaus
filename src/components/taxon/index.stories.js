import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import FetchContainer from '../../FetchContainer'

import { Paper } from 'material-ui'
import TaxonList from './TaxonList'
import TaxonGallery from './TaxonGallery'
import taxon from './testTaxon'

function mock(description, url, callback) {
  console.log(description, url)
  callback(taxon)
}

storiesOf('Taxon', module)
  .add('List', () => (
    <FetchContainer fetchJson={mock}>
      <TaxonList taxonId={59764} onPickTaxon={action('select')} />
    </FetchContainer>
  ))
  .add('Gallery', () => (
    <FetchContainer fetchJson={mock}>
      <TaxonGallery taxonId={59764} onPickTaxon={action('select')} />
    </FetchContainer>
  ))
