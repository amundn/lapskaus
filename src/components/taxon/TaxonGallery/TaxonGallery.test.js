import React from 'react'
import renderer from 'react-test-renderer'
import TaxonGallery from './TaxonGallery'
import TaxonGrid from './TaxonGrid'
import TaxonTile from './TaxonTile'
import FetchContainer from '../../../FetchContainer'
import taxon from '../testTaxon'

test('Taxon gallery', () => {
  const component = renderer.create(<TaxonGallery tree={taxon} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
