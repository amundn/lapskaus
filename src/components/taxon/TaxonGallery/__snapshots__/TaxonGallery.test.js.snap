// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Taxon gallery 1`] = `
<div>
  <ul
    className="MuiList-root-1 MuiList-padding-2"
  >
    <div>
      <li
        className="MuiButtonBase-root-15 MuiListItem-root-5 MuiListItem-gutters-12 MuiListItem-button-13 MuiListItem-default-8"
        disabled={false}
        id={0}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        role="button"
        style={
          Object {
            "xmarginLeft": 0,
          }
        }
        tabIndex={0}
      >
        <div
          className="MuiListItemText-root-17"
        >
          <h3
            className="MuiTypography-root-22 MuiTypography-subheading-29 MuiListItemText-text-20"
          >
            <span
              style={
                Object {
                  "textTransform": "capitalize",
                }
              }
            >
              Life (Life)
            </span>
          </h3>
        </div>
        <span
          className="MuiTouchRipple-root-44"
        />
      </li>
      <li
        className="MuiButtonBase-root-15 MuiListItem-root-5 MuiListItem-gutters-12 MuiListItem-button-13 MuiListItem-default-8"
        disabled={false}
        id={1131}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        role="button"
        style={
          Object {
            "xmarginLeft": 0,
          }
        }
        tabIndex={0}
      >
        <div
          className="MuiAvatar-root-51"
        >
          <img
            alt={undefined}
            className="MuiAvatar-img-53"
            sizes={undefined}
            src="//nodeyoda.westeurope.cloudapp.azure.com/taxonPhoto/Plantae.jpg"
            srcSet={undefined}
          />
        </div>
        <div
          className="MuiListItemText-root-17"
        >
          <h3
            className="MuiTypography-root-22 MuiTypography-subheading-29 MuiListItemText-text-20"
          >
            <span
              style={
                Object {
                  "textTransform": "capitalize",
                }
              }
            >
              planteriket (Plantae)
            </span>
          </h3>
        </div>
        <span
          className="MuiTouchRipple-root-44"
        />
      </li>
    </div>
  </ul>
  <div
    className="TaxonGrid-container-54"
  >
    <ul
      className="MuiGridList-root-56 TaxonGrid-gridList-55"
      style={
        Object {
          "margin": -2,
        }
      }
    >
      <li
        className="MuiGridListTile-root-57"
        style={
          Object {
            "height": "auto",
            "padding": 2,
            "width": "100%",
          }
        }
      >
        <div
          className="MuiGridListTile-tile-58"
        >
          <div
            className="MuiListSubheader-root-61 MuiListSubheader-sticky-65"
          >
            dekkfroete blomsterplanter (Magnoliophyta)
          </div>
        </div>
      </li>
      <li
        className="MuiGridListTile-root-57"
        onClick={[Function]}
        style={
          Object {
            "margin": "2px",
          }
        }
      >
        <div
          className="MuiGridListTile-tile-58"
        >
          <div
            style={
              Object {
                "backgroundImage": "url(//nodeyoda.westeurope.cloudapp.azure.com/taxonphoto/Monocots.jpg)",
                "backgroundPosition": "center center",
                "backgroundRepeat": "no-repeat",
                "backgroundSize": "cover",
                "height": "379px",
                "position": "relative",
                "width": "379px",
              }
            }
          />
          <div
            className="MuiGridListTileBar-root-69 TaxonTile-titleBar-67 MuiGridListTileBar-rootTop-71 MuiGridListTileBar-rootWithSubtitle-72"
          >
            <div
              className="MuiGridListTileBar-titleWrap-73 MuiGridListTileBar-titleWrapActionRight-75"
            >
              <div
                className="MuiGridListTileBar-title-76 TaxonTile-title-66"
              >
                <span
                  style={
                    Object {
                      "textTransform": "capitalize",
                    }
                  }
                >
                  enfrobladete blomsterplanter
                </span>
              </div>
              <div
                className="MuiGridListTileBar-subtitle-77"
              >
                <b>
                  Monocots
                </b>
              </div>
            </div>
            <div
              className=""
            >
              <button
                className="MuiButtonBase-root-15 MuiIconButton-root-80"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label-86"
                >
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root-89"
                    color="white"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 4H5c-1.11 0-2 .9-2 2v12c0 1.1.89 2 2 2h4v-2H5V8h14v10h-4v2h4c1.1 0 2-.9 2-2V6c0-1.1-.89-2-2-2zm-7 6l-4 4h3v6h2v-6h3l-4-4z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-44"
                />
              </button>
            </div>
          </div>
        </div>
      </li>
      <li
        className="MuiGridListTile-root-57"
        onClick={[Function]}
        style={
          Object {
            "margin": "2px",
          }
        }
      >
        <div
          className="MuiGridListTile-tile-58"
        >
          <div
            style={
              Object {
                "backgroundImage": "url(//nodeyoda.westeurope.cloudapp.azure.com/taxonphoto/Eudicots.jpg)",
                "backgroundPosition": "center center",
                "backgroundRepeat": "no-repeat",
                "backgroundSize": "cover",
                "height": "379px",
                "position": "relative",
                "width": "379px",
              }
            }
          />
          <div
            className="MuiGridListTileBar-root-69 TaxonTile-titleBar-67 MuiGridListTileBar-rootTop-71 MuiGridListTileBar-rootWithSubtitle-72"
          >
            <div
              className="MuiGridListTileBar-titleWrap-73 MuiGridListTileBar-titleWrapActionRight-75"
            >
              <div
                className="MuiGridListTileBar-title-76 TaxonTile-title-66"
              >
                <span
                  style={
                    Object {
                      "textTransform": "capitalize",
                    }
                  }
                >
                  tofrobladete blomsterplanter
                </span>
              </div>
              <div
                className="MuiGridListTileBar-subtitle-77"
              >
                <b>
                  Eudicots
                </b>
              </div>
            </div>
            <div
              className=""
            >
              <button
                className="MuiButtonBase-root-15 MuiIconButton-root-80"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label-86"
                >
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root-89"
                    color="white"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 4H5c-1.11 0-2 .9-2 2v12c0 1.1.89 2 2 2h4v-2H5V8h14v10h-4v2h4c1.1 0 2-.9 2-2V6c0-1.1-.89-2-2-2zm-7 6l-4 4h3v6h2v-6h3l-4-4z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root-44"
                />
              </button>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
